lib/RestructureCFG/SelectScopePass.cpp

```diff
@@ -0,0 +1,294 @@
+Logger<> ScopeLog("select-scope");
+
+static SetVector<BasicBlock *> getConditionalSuccessors(BasicBlock *N) {
+  // We employ a `SetVector` so that we do not take into account
+  // multiplicity for edges out of a conditional
+  SetVector<BasicBlock *> ConditionalSuccessors;
+  for (BasicBlock *Successor : children<Scope<BasicBlock *>>(N)) {
+    ConditionalSuccessors.insert(Successor);
+  }
+
+  return ConditionalSuccessors;
+}
+
+static SetVector<BasicBlock *> getPredecessors(BasicBlock *N) {
+  // It is important that we use a `SetVector` here in order to
+  // deduplicate the successors outputted by the `llvm::children` range
+  // iterator
+  SetVector<BasicBlock *> Predecessors;
+  for (auto *Predecessor : children<Inverse<Scope<BasicBlock *>>>(N)) {
+    revng_log(ScopeLog, "  Predecessor: " + Predecessor->getName().str());
+    Predecessors.insert(Predecessor);
+  }
+
+  return Predecessors;
+}
+
+static SmallVector<BasicBlock *> getNodesToProcess(BasicBlock *Conditional,
+                                                   BasicBlock *PostDominator) {
+  // We exploit the `Visited` set, by passing it to
+  // `ReversePostOrderTraversalExt`, in order to stop the visit at the
+  // `PostDominator`
+  std::set<BasicBlock *> Visited;
+  Visited.insert(PostDominator);
+
+  // We collect all the nodes between the `Conditional` and its
+  // immediate postdominator, by using the `ReversePostOrderTraversalExt`
+  SmallVector<BasicBlock *> NodesToProcess;
+  for (BasicBlock *RPONode :
+       ReversePostOrderTraversalExt<Scope<BasicBlock *>>(Conditional,
+                                                         Visited)) {
+    NodesToProcess.push_back(RPONode);
+  }
+
+  // From the collected nodes, we need to remove the first node, which
+  // corresponds to the `Conditional`, which should not be processed in this
+  // round
+  revng_assert(NodesToProcess.front() == Conditional);
+  NodesToProcess.erase(NodesToProcess.begin());
+
+  return NodesToProcess;
+}
+
+static std::map<BasicBlock *, size_t>
+initializeScopeMap(SetVector<BasicBlock *> &ConditionalSuccessors) {
+  std::map<BasicBlock *, size_t> ScopeMap;
+
+  // We initialize the `ScopeMap` for each `ConditionalSuccessor`
+  for (const auto &[Index, ConditionalSuccessor] :
+       enumerate(ConditionalSuccessors)) {
+    if (not ScopeMap.contains(ConditionalSuccessor)) {
+      ScopeMap[ConditionalSuccessor] = Index;
+    }
+  }
+  return ScopeMap;
+}
+
+/// Helper function used to elect the `ScopeID` for a node, electing the
+/// `ScopeID` for which we have the highest number of predecessors having such
+/// `ScopeID`
+static size_t electMaxScopeID(SetVector<BasicBlock *> &Predecessors,
+                              std::map<BasicBlock *, size_t> &ScopeMap) {
+  // In the map, we store the how frequent a `ScopeID` is in the
+  // predecessors of `Candidate`, in order to elect the one with the most
+  // common `ScopeID`
+  std::map<size_t, size_t> PredecessorsScopeCounter;
```

Comment:
Use a `DenseMap` here.

Reply:
Changed to `DenseMap` here https://github.com/revng/revng/blob/9208e1d0d7ecf56a86e3f44debb589c8eded149c/lib/RestructureCFG/InlineDivergentScopesPass.cpp#L237